src/main/java/org/elasticsearch/common/util/concurrent/ReleasableLock.java

```diff
@@ -30,17 +30,47 @@
 public class ReleasableLock implements Releasable {
     private final Lock lock;
 
+    private final ThreadLocal<Boolean> holdingThreads;
+
     public ReleasableLock(Lock lock) {
         this.lock = lock;
+        boolean useHoldingThreads = false;
+        assert (useHoldingThreads = true);
+        if (useHoldingThreads) {
+            holdingThreads = new ThreadLocal<>();
+        } else {
+            holdingThreads = null;
+        }
     }
 
     @Override
     public void close() {
         lock.unlock();
+        assert removeCurrentThread();
     }
 
+
     public ReleasableLock acquire() throws EngineException {
         lock.lock();
+        assert addCurrentThread();
         return this;
     }
+
+    private boolean addCurrentThread() {
+        holdingThreads.set(true);
+        return true;
+    }
+
+    private boolean removeCurrentThread() {
+        holdingThreads.remove();
+        return true;
+    }
+
+    public Boolean isHeldByCurrentThread() {
+        if (holdingThreads == null) {
+            return null;
```

Comment:
I feel weird about returning null here, if holdingThreads is null, can we just return false? Or are we signaling something special (asserts aren't enabled) with this?